presto-password-authenticators/src/main/java/io/prestosql/plugin/password/filebasedauthenticator/User.java

```diff
@@ -0,0 +1,174 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.password.filebasedauthenticator;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import io.airlift.log.Logger;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.StandardErrorCode;
+
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkAndGetHashing;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkBCryptPassword;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkPBKDF2Password;
+import static java.util.Objects.requireNonNull;
+
+public class User
+{
+    private final String name;
+    private final String password;
+    private static final Logger log = Logger.get(User.class);
+    private Optional<HashingAlgorithm> hashingAlgo = Optional.empty();
+    private Cache<PasswordStore, Boolean> authCacheStore;
+
+    public User(String name, String password, Cache<PasswordStore, Boolean> bcryptCache, Cache<PasswordStore, Boolean> pbkdf2Cache, FileBasedAuthenticationConfig config) throws HashedPasswordException
+    {
+        this.name = requireNonNull(name, "User name is empty");
+        requireNonNull(password, "User name is empty");
+
+        this.password = password;
+
+        hashingAlgo = Optional.of(checkAndGetHashing(password, config.bcryptMinCost, config.pbkdf2MinIterations));
+        switch (hashingAlgo.get()) {
+            case BCRYPT:
+                authCacheStore = bcryptCache;
+                break;
+            case PBKDF2:
+                authCacheStore = pbkdf2Cache;
+                break;
+            default:
+                throw new PrestoException(StandardErrorCode.CONFIGURATION_INVALID, "Hashing algorithm of password, stored in the {user:password} file cannot be determined!");
+        }
+    }
+
+    public class PasswordStore
+    {
+        private final String inputUsername;
+        private final String inputPassword;
+
+        public PasswordStore(String username, String password)
+        {
+            this.inputUsername = username;
+            this.inputPassword = password;
+        }
+
+        public String getInputUsername()
+        {
+            return this.inputUsername;
+        }
+
+        public String getInputPassword()
+        {
+            return this.inputPassword;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return inputUsername.hashCode() * 101 + (inputPassword == null ? 0 : inputPassword.hashCode());
```

Comment:
Use `Objects.equals` and `Objects.hash`. See `io.prestosql.sql.tree.Window` for an example.

Reply:
Done